frontend/ui/widget/menu.lua

```diff
@@ -281,34 +286,116 @@ function MenuItem:init()
-                local ellipsis_size = RenderText:sizeUtf8Text(0, self.content_width,
-                    Font:getFace(self.font, self.font_size), "…", true, self.bold).x
-                local removed_char_width= 0
-                while removed_char_width < ellipsis_size  do
-                    -- the width of each char has already been calculated by TextBoxWidget
-                    removed_char_width = removed_char_width + item_name:getCharWidth(offset)
-                    offset = offset - 1
+            flag_add_ellipsis = true
+            num_lines = item_name:getAllLineCount()
+            if num_lines == 1 then   -- widget doesn't fit and we have only one line of text
+                break
+            end
+            -- remove last line and try again to fit
+            offset = item_name.vertical_string_list[num_lines].offset - 1
+            self.text = table.concat(item_name.charlist, '', 1, offset)
+        end
+
+        -- second: decrease font size (we have now only one line)
+        while true and not flag_fit do
+            -- Free previously made widgets to avoid memory leaks
+            if item_name then
+                item_name:free()
+            end
+            self.font_size = self.font_size - 1
+            item_name = TextBoxWidget:new {
+                text = self.text,
+                face = Font:getFace(self.font, self.font_size),
+                width = self.content_width - mandatory_w - state_button_width - text_mandatory_padding,
+                alignment = "left",
+                bold = self.bold,
+                fgcolor = self.dim and Blitbuffer.COLOR_DARK_GRAY or nil,
+            }
+            local height = item_name:getSize().h
+            if height < max_item_height then -- we fit !
+                offset = #item_name.charlist
+                break
+            end
+
+            -- if text is too height with that really small font we don't show text at all
+            -- this shouldn't happen
+            if self.font_size <= 8 then
+                item_name = TextBoxWidget:new {
+                    text = "…",
+                    face = Font:getFace(self.font, self.font_size),
+                    width = self.content_width - mandatory_w - state_button_width - text_mandatory_padding,
+                    alignment = "left",
+                    bold = self.bold,
+                    fgcolor = self.dim and Blitbuffer.COLOR_DARK_GRAY or nil,
+                }
+                flag_add_ellipsis = false
+                break
+            end
+        end
+        -- add ellipsis when text was truncated
+        if flag_add_ellipsis then
+            local text_last_line
+            -- when lines is more than 1 we see only for last visible line
+            if num_lines > 1 then
+                local offset_prev = item_name_orig.vertical_string_list[num_lines - 1].offset
+                text_last_line = table.concat(item_name_orig.charlist, '', offset_prev, offset)
```

Comment:
One of the things I intend to do is to move that kind of stuff into TextBoxWidget.
It already has a "private" `TextBoxWidget:_getLineText(vertical_string)` which does what you do, so I guess it just have to have a "public" `TextBoxWidget:getLineText(line_num)` to give you that.
So, not asking you to do it :) This is fine for now.

Reply:
Random single quotes instead of double.